### ISSUE TITLE
Why not use torch.nn.utils.clip in multiprocessing_trainer.py?

### ISSUE THREAD

**Issue by weilheim** [facebookresearch/fairseq-py#18]
I don't understand why you write your own functions  "_clip_grads_" and "_flatten_grads_" to clip the gradient instead of using torch.nn.utils.clip. I have read the source code of  torch.nn.utils.clip and I think it does exactly the same thing as your functions.
Is it because that your functions is faster than the original one? Could anyone explain the reason?
Thanks!

**Comment by myleott** [comment]
It's strictly to improve performance. We flatten the grads into a contiguous Tensor so that we only have to call all_reduce once. This is similar to how PyTorch's DistributedDataParallel does it [1]. We then defined our own clip_grads function since, once we've flattened the Tensor, there's little point iterating over the individual parameters to compute the norm.

[1] https://github.com/pytorch/pytorch/blob/6743d5951397eaf0bd454c941af00d2d0fae096d/torch/nn/parallel/distributed.py#L298
